### PR TITLE
[Refactor] Create BaseLightningModule without nn.Module + small modifs

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces the BaseLightningModule as a base for LightningModule.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
what is the motivation, can you give some context?
also seems to be as just moving to another file?

**Comment by tchaton** [comment]
> what is the motivation, can you give some context?
> also seems to be as just moving to another file?

**Comment by ananthsub** [comment]
as mentioned on chat, I think this is the wrong abstraction/fit. What's the difference between the RootLightningModule and the LightningModule? Could you describe the motivation? Is there an issue which has more context?

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -451,7 +129,7 @@ def __check_not_nested(value: dict, name: str) -> None:
         return value
 
     @staticmethod
-    def __check_allowed(v: Any, name: str, value: Any) -> None:
+    def _check_allowed(v: Any, name: str, value: Any) -> None:
```
Comment (ethanwharris) [inline pytorch_lightning/core/lightning.py:132]: This is duplicated in `RootLightningModule`, is that right?

File: pytorch_lightning/trainer/supporters.py [hunk 2]
```diff
@@ -377,6 +377,11 @@ def sampler(self) -> Union[Iterable, Sequence, Mapping]:
         """Return a collections of samplers extracting from loaders."""
         return apply_to_collection(self.loaders, (DataLoader, IterableDataset), getattr, 'sampler', None)
 
+    @property
+    def batch_size(self) -> Union[Iterable, Sequence, Mapping]:
+        """Return a collections of samplers extracting from loaders."""
```
Comment (ethanwharris) [inline pytorch_lightning/trainer/supporters.py:382]:
```suggestion
        """Return a collection of batch sizes extracted from loaders."""
```

**Review by ethanwharris — APPROVED** [review]
LGTM, couple minor comments :smiley:

File: pytorch_lightning/core/base_lightning.py [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+from pytorch_lightning.utilities.warnings import WarningCache
+
+if TYPE_CHECKING:
+    from pytorch_lightning.trainer.connectors.logger_connector.result import Result
+
+warning_cache = WarningCache()
+log = logging.getLogger(__name__)
+
+
+class RootLightningModule(ABC, ModelHooks, DataHooks, CheckpointHooks):
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/core/base_lightning.py:47]: Changed for `BaseLightningModule`.

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:132]: Yes, removed. Good catch !

File: pytorch_lightning/trainer/supporters.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py:382]: Done !

**Comment by tchaton** [comment]
> as mentioned on chat, I think this is the wrong abstraction/fit. What's the difference between the RootLightningModule and the LightningModule? Could you describe the motivation? Is there an issue which has more context?

Hey @ananthsub. You are right ! 
This PR is just the first step toward building up the right abstraction for the LightningModule and extending its functionalities.
And we will investigate your suggested API. 

Best,
T.C

File: pytorch_lightning/core/hooks.py [hunk 4]
```diff
@@ -651,7 +651,7 @@ def on_test_dataloader(self) -> None:
     def on_predict_dataloader(self) -> None:
         """Called before requesting the predict dataloader."""
 
-    def transfer_batch_to_device(self, batch: Any, device: torch.device, dataloader_idx: int) -> Any:
+    def transfer_batch_to_device(self, batch: Any, device: torch.device, dataloader_idx: int = None) -> Any:
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:654]:
```suggestion
    def transfer_batch_to_device(self, batch: Any, device: torch.device, dataloader_idx: Optional[int] = None) -> Any:
```

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -92,6 +92,9 @@ class Trainer(
     DeprecatedTrainerAttributes,
 ):
 
+    accelerator_connector_cls = AcceleratorConnector
+    checkpoint_connector_cls = CheckpointConnector
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:96]: So it looks like in this "other library" you will set a different class on the trainer right, one that you subclassed?
In the future, I can see that you will do more of these changes, like other connectors, loops, and so on. Is that the plan?

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:96]: I'm very interested in this too since our guidance so far has been to not extend the trainer class.

Comment (kaushikb11) [reply pytorch_lightning/trainer/trainer.py:96]: There's an issue, with this design, the arguments for `AcceleratorConnector` and `CheckpointConnector` of PyTorch Lightning would need to match the other library's, which could cause issues down the road.

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:96]: I don't think making connectors part of the public API makes sense to user. Users should only need to pass the constructed accelerator or checkpoint callback objects.

IMO supporting this other library should not sacrifice or make the existing APIs more complex, especially if it's  experimental.

File: pytorch_lightning/utilities/xla_device.py [hunk 6]
```diff
@@ -71,6 +71,10 @@ def _is_device_tpu() -> bool:
         # we would have to use `torch_xla.distributed.xla_dist` for
         # multiple VMs and TPU_CONFIG won't be available, running
         # `xm.get_xla_supported_devices("TPU")` won't be possible.
+
+        # Temporary to support tpu+vm arch
+        if "XRT_TPU_CONFIG" not in os.environ:
+            return False
```
Comment (ananthsub) [inline pytorch_lightning/utilities/xla_device.py:77]: won't this change existing behavior for TPU training?

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -57,15 +57,14 @@
 log = logging.getLogger(__name__)
 
 
-class LightningModule(
+class BaseLightningModule(
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:60]: Is this meant to be exposed publicly? If not, could we prefix it like this? 
```suggestion
class _BaseLightningModule(
```

assuming the other library would have a different top-level module that would be extended.

File: pytorch_lightning/utilities/xla_device.py (hunk 6 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/xla_device.py:77]: Nope, as the env variable is required by PyTorch XLA

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/trainer.py:96]: > I don't think making connectors part of the public API makes sense to user. Users should only need to pass the constructed accelerator or checkpoint callback objects.

Agreed.

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py]: Done !

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Great points ! Removed it entirely.

File: pytorch_lightning/utilities/xla_device.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/utilities/xla_device.py:77]: @kaushikb11 @tchaton n00b question: why does the comment say its temporary then?

File: pytorch_lightning/accelerators/accelerator.py [hunk 8]
```diff
@@ -114,10 +114,11 @@ def pre_dispatch(self, trainer: 'pl.Trainer') -> None:
     def _move_optimizer_state(self) -> None:
         """ Moves the state of the optimizers to the GPU if needed. """
         for opt in self.optimizers:
-            state: DefaultDict = defaultdict(dict)
-            for p, v in opt.state.items():
-                state[p] = apply_to_collection(v, torch.Tensor, move_data_to_device, self.root_device)
-            opt.state = state
+            if isinstance(opt.state, DefaultDict):
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:117]: looks like a bugfix here, maybe worth a changelog update?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -57,15 +57,14 @@
 log = logging.getLogger(__name__)
 
 
-class LightningModule(
+class _BaseLightningModule(
     ABC,
     DeviceDtypeModuleMixin,
```
Comment (justusschock) [inline pytorch_lightning/core/lightning.py:62]: This one also inherits from Module, not sure if we have to have it here, but currently it does.

Also: if we say we don't want this one to inherit from nn.Module, we have to reimplement everything that is delegated back to nn.Module here again for the other lib

File: pytorch_lightning/trainer/supporters.py [hunk 10]
```diff
@@ -377,6 +377,11 @@ def sampler(self) -> Union[Iterable, Sequence, Mapping]:
         """Return a collections of samplers extracting from loaders."""
         return apply_to_collection(self.loaders, (DataLoader, IterableDataset), getattr, 'sampler', None)
 
+    @property
+    def batch_size(self) -> Union[Iterable, Sequence, Mapping]:
```
Comment (justusschock) [inline pytorch_lightning/trainer/supporters.py:381]: what is this needed for? This should be a separate PR if we really need it (don't see a reason for that right now)

**Review by justusschock — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/supporters.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py:381]: Fine.
